include/arch/arm/aarch32/exc.h

```diff
@@ -91,7 +92,28 @@ struct __esf {
 #endif
 };
 
-typedef struct __esf z_arch_esf_t;
+/* The registers not preserved by hardware on exception entry */
+struct __callee_saved_esf {
+	uint32_t r4;
+	uint32_t r5;
+	uint32_t r6;
+	uint32_t r7;
+	uint32_t r8;
+	uint32_t r9;
+	uint32_t r10;
+	uint32_t r11;
+};
+
+struct __full_esf {
```

Comment:
I am not sure this is the right thing to do.
The callee-saved registers are not part of the the exception stack frame, and adding the container here is confusing.

In my opinion, the best thing is to keep the callee-saved registers information outside the exception stack frame.
The values of the registers are _absolutely_ not needed by the kernel to determine the root cause of the errors; the values should not at all be used in fault.c

That said, I would propose:
- use the existing callee-saved register structure, defined for Cortex-m:
```
struct _callee_saved {
	uint32_t v1;  /* r4 */
	uint32_t v2;  /* r5 */
	uint32_t v3;  /* r6 */
	uint32_t v4;  /* r7 */
	uint32_t v5;  /* r8 */
	uint32_t v6;  /* r9 */
	uint32_t v7;  /* r10 */
	uint32_t v8;  /* r11 */
	uint32_t psp; /* r13 */
};

typedef struct _callee_saved _callee_saved_t;
```

- In fault.S store and pass a pointer to such a struct in z_arm_fault
- In fault.c do nothing, but _transparently_ pass this information in fatal.c

Reply:
Thanks for the thoughts @ioannisg! One goal I'm trying to solve for here is passing the information into the weak function `k_sys_fatal_error_handler`. This allows an end user to implement a handler that saves off the the stack & register state needed to recover a full backtrace after a crash has taken place. If the callee saved state is removed from `z_arch_esf_t`, how would you propose passing that state so an end user can collect it?

Reply:
Ah, I see, so it is not sufficient for the fatal.c to process the information, right? I didn't think about it.
In that case, I guess we need to include the callee-saved regs into the z_arch_esf_t structure. Check if this can be done by adding a struct _callee_saved pointer so we can reuse the already defined structure.

Please, investigate, if this can be done in a clean and transparent way, that is
- fault.S to include a pointer to a structure that holds the callee-saved regs as we discussed above, when calling z_arm_fault
- fault.c to _copy_ the callee-saved regs into the stacked copy of z_arch_esf_t right before calling z_arm_fatal_error (ideally, no more changes should be needed)

Could this work?